server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -146,7 +146,7 @@ public TransportSearchAction(ThreadPool threadPool,
         final Set<String> indicesAndAliases = indexNameExpressionResolver.resolveExpressions(clusterState, request.indices());
         for (Index index : concreteIndices) {
             clusterState.blocks().indexBlockedRaiseException(ClusterBlockLevel.READ, index.getName());
-            AliasFilter aliasFilter = searchService.buildAliasFilter(clusterState, index.getName(), indicesAndAliases);
+            AliasFilter aliasFilter = searchService.buildAliasFilter(clusterState, index.getName(), indicesAndAliases, false);
```

Comment:
From my reading of the code, I understand that this code change ALMOST covers the `ShardSearchRequest` authz improvement case @ywangd handled in https://github.com/elastic/elasticsearch/pull/78327 . 
Specifically, the `ShardSearchRequest` should contain the alias and/or data stream alias that is used in the original search request to refer to the shard (so security can authorize on the alias name).

I think there might be two problems here (from an authz point of view)
* We need to decide if we support aliases pointing to backing indices of data streams; if we do, the `originalIndices` field must contain both kinds of aliases - fun!
* In authorization, we need to account for the fact that the `originalIndices` only contains aliases (as it currently stands and assuming my reading of the code is correct) but not also the index name (we can fish it out from the shard id)

What's you reckon @ywangd ?

Reply:
> We need to decide if we support aliases pointing to backing indices of data streams; if we do, the originalIndices field must contain both kinds of aliases - fun!

That's what this PR does.

> In authorization, we need to account for the fact that the originalIndices only contains aliases (as it currently stands and assuming my reading of the code is correct) but not also the index name (we can fish it out from the shard id)

OriginalIndices can contain the original index name or the alias, not both. That's the case today since we want to distinguish accessing an index from an alias. Do you have an example in mind that this PR changes ?

Reply:
> OriginalIndices can contain the original index name or the alias, not both. That's the case today since we want to distinguish accessing an index from an alias. Do you have an example in mind that this PR changes ?

The authorization code needs all relevant names (including alias, data stream, index) for the concrete index as long as they are part of the original requested indices. For example, if concrete index `my-index1` has an alias `my-alias1`, for a search request like `GET my-*/_search`, we need both `my-index1` and `my-alias1` to be there in the `ShardSearchRequest`. 

This is mostly needed for handling DLS/FLS. Currently if there are multiple DLS/FLS defined for the same underlying concrete indices, they are merged and the end result is an union of all DLS/FLS. That is, the following index privileges:
```
{
  "indices": [
    {
      "names": [ "my-index1" ],
      "privileges": [ "read" ],
      "query": {"term": {"value": 42}}
    },
    {
      "names": ["my-alias1"],
      "privileges": ["read"],
      "query": {"term": {"value": 1}}
    }
  ]
}
```
means that the user has access to `my-index1` for any documents that have either value of `42` or `1`. 

Therefore the Shard level request needs to carry both of the names `my-index1` and `my-alias1` to the remote node. Otherwise the result will not be correct (based on the correctness defined as today's code).

The same logic applies to data streams and its backing indices as well, or data stream aliases and the backings. In summary, any names that are relevant to the concrete index and also appear (either directly or as an expansion) in the original request need to be carried by the shard/index level requests.

Reply:
We talked offline and I pushed https://github.com/elastic/elasticsearch/pull/78508/commits/82beb936499284e32b5f207f5f28bce21be6ab58 to preserve the original index name when needed. This is not necessary for data streams because we propagate authorizations to the backing indices automatically. 

Cross cluster search implements [the wrong logic](https://github.com/elastic/elasticsearch/pull/24502) since 5.5 though. This PR does not try to fix this bug, I am going to open a separate PR/issue to fix this long standing bug.